File: 20221125_Friday/20221125.rb
# Leetcode: 907. Sum of Subarray Minimums.
# https://leetcode.com/problems/sum-of-subarray-minimums/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 182 ms, faster than 100.00% of Ruby online submissions for Sum of Subarray Minimums.
# Memory Usage: 215.7 MB, less than 71.43% of Ruby online submissions for Sum of Subarray Minimums.
# 2022.11.25 Daily Challenge.
# @param {Integer[]} arr
# @return {Integer}
def sum_subarray_mins(arr)
  # Based on:
  # https://leetcode.com/problems/sum-of-subarray-minimums/solution/
  mod = 10**9 + 7
  stack = []
  sum = 0
  (0..arr.size).each do |i|
    # when `i` reaches the array lenght, it is an indication that
    # all the elements have been processed, and the remaining
    # elements in the stack should no be popped out.
    while (! stack.empty?) && (arr.size == i || arr[stack.last] >= arr[i])
      # Notice the sign ">=". This ensures that no contribution
      # is counted twice. right_boundary takes equal or smaller
      # elements into account while left_boundary takes only the
      # strictly smaller elements into account.
      mid = stack.pop()
      left_boundary = stack.empty? ? -1 : stack.last
      right_boundary = i

      # Count of subarrays where mid is the minimum elemnt.
      count = (mid - left_boundary) * (right_boundary - mid)
      sum += count * arr[mid]
    end
    stack.append(i)
  end
  return sum % mod
end

File: 20221125_Friday/20221125.go
// Leetcode: 907. Sum of Subarray Minimums.
// https://leetcode.com/problems/sum-of-subarray-minimums/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// 2022.11.25 Daily Challenge.
import "fmt"
var p  = fmt.Println
var mod int = 1000000007
type stack struct{
  data []int
}
func (s * stack) init(){
  s.data = make([]int,0)
}
func (s * stack) pop() int{
  ans := s.data[len(s.data)-1]
  s.data = s.data[0:len(s.data)-1]  // Fix. 
  return ans
}
func (s * stack) push(x int){
  s.data = append(s.data, x)
}
func (s * stack) empty() bool{
  return len(s.data) == 0
}
func (s * stack) last() int {
  return s.data[len(s.data)-1]
}

func sumSubarrayMins(arr []int) int {
  // Based on:
  // https://leetcode.com/problems/sum-of-subarray-minimums/solution/
  sum := 0
  var s stack
  s.init()
  for i := 0 ; i <= len(arr); i += 1{
    for ; (!s.empty()) && ( i == len(arr) || arr[s.last()] >= arr[i]);  {
      mid := s.pop()
      lb := - 1
      if(!s.empty()) {lb = s.last()}
      rb := i
      count := (mid - lb) * (rb - mid)
      sum +=  (count * arr[mid])%mod
      sum = sum % mod // Fix.
    }
    s.push(i)
  }
  return sum
}
